src/google/cloud/ndb/context.py

```diff
@@ -77,6 +78,24 @@ def get_context():
     raise exceptions.ContextError()
 
 
+class _Cache(collections.UserDict):
+    """An in-memory entity cache.
+
+       This cache verifies the fetched entity has the correct key before
+       returning a result, in order to handle cases where the entity's key was
+       modified but the cache's key was not updated."""
+
+    def get_and_validate(self, key):
+        """Verify that the entity's key has not changed since it was added
+           to the cache. If it has changed, consider this a cache miss.
+           See issue 13.  http://goo.gl/jxjOP"""
+        entity = self.data[key]  # May be None, meaning "doesn't exist".
+        if entity is None or entity._key == key:
+            return entity
+        else:
+            raise KeyError(key)
```

Comment:
del self[key]  ???

Reply:
Okay, added (it had to be del self.data[key] to avoid a recursive loop as del fires getitem)